snowfakery/template_funcs.py

```diff
@@ -130,14 +130,26 @@ def reference(self, x: Union[ObjectRow, str]):
             if hasattr(x, "id"):  # reference to an object with an id
                 target = x
             elif isinstance(x, str):  # name of an object
-                obj = self.context.field_vars().get(x)
-                if not obj:
+                # allows dotted paths
+                parts = x.split(".")
+                target = self.context.field_vars().get(parts.pop(0))
+
+                for part in parts:
+                    try:
+                        target = getattr(target, part)
+                    except AttributeError:
+                        raise DataGenError(
+                            f"Expression cannot be evaluated `{x}``. Problem before `{part}`: {target}",
+                            None,
+                            None,
```

Comment:
Should DataGenError default to None for the 2nd and 3rd args so it's not necessary to spell them out?

Reply:
Maybe. I wanted to reduce the likelihood of forgetting to supply them when they were available but the deeper you get into the interpreter, the less likely it is that they are available.

Reply:
Isn't it the sort of thing where the outer layers of the interpreter should keep track of the context and be responsible for annotating any exceptions raised by the inner layers?

Making them required hasn't solved the problem of forgetting them: https://github.com/SFDO-Tooling/Snowfakery/pull/161#discussion_r508553754

Reply:
#273